File: deque/MyCircularDeque.java

/*
641. Design Circular Deque
Medium

https://leetcode.com/problems/design-circular-deque/

Runtime: 4 ms, faster than 98.93% of Java online submissions for Design Circular Deque.
Memory Usage: 40 MB, less than 5.07% of Java online submissions for Design Circular Deque.

Design your implementation of the circular double-ended queue (deque).

Your implementation should support following operations:

    MyCircularDeque(k): Constructor, set the size of the deque to be k.
    insertFront(): Adds an item at the front of Deque. Return true if the operation is successful.
    insertLast(): Adds an item at the rear of Deque. Return true if the operation is successful.
    deleteFront(): Deletes an item from the front of Deque. Return true if the operation is successful.
    deleteLast(): Deletes an item from the rear of Deque. Return true if the operation is successful.
    getFront(): Gets the front item from the Deque. If the deque is empty, return -1.
    getRear(): Gets the last item from Deque. If the deque is empty, return -1.
    isEmpty(): Checks whether Deque is empty or not. 
    isFull(): Checks whether Deque is full or not.

 

Example:

MyCircularDeque circularDeque = new MycircularDeque(3); // set the size to be 3
circularDeque.insertLast(1);			// return true
circularDeque.insertLast(2);			// return true
circularDeque.insertFront(3);			// return true
circularDeque.insertFront(4);			// return false, the queue is full
circularDeque.getRear();  			// return 2
circularDeque.isFull();				// return true
circularDeque.deleteLast();			// return true
circularDeque.insertFront(4);			// return true
circularDeque.getFront();			// return 4

 

Note:

    All values will be in the range of [0, 1000].
    The number of operations will be in the range of [1, 1000].
    Please do not use the built-in Deque library.

Accepted
21,119
Submissions
38,779

*/
class MyCircularDeque {
    
    class Node {
        int value;
        Node prev, next;
        Node(int value){
            this.value = value;
        }
        Node(int value, Node prev, Node next){
            this.value = value;
            this.prev = prev;
            this.next = next;
        }
    }

    /** Initialize your data structure here. Set the size of the deque to be k. */
    int count, length;
    Node head, tail;
    public MyCircularDeque(int k) {
        this.count = 0;
        this.length = k;
    }
    
    /** Adds an item at the front of Deque. Return true if the operation is successful. */
    public boolean insertFront(int value) {
        if(this.isFull())
            return false;
        if(this.isEmpty()){
            this.head = this.tail = new Node(value);
            this.tail.next = this.head;
            this.head.prev = this.tail;
        }
        else{
            this.head = new Node(value, this.tail, this.head);
            this.tail.next = this.head;
            this.head.next.prev = this.head;
        }
        this.count++;
        return true;
    }
    
    /** Adds an item at the rear of Deque. Return true if the operation is successful. */
    public boolean insertLast(int value) {
        if(this.isFull())
            return false;
        if(this.isEmpty()){
            this.head = this.tail = new Node(value);
            this.tail.next = this.head;
            this.head.prev = this.tail;
        }
        else{
            this.tail.next = new Node(value, this.tail, this.head);
            this.tail = this.tail.next;
            this.head.prev = this.tail;
        }
        this.count++;
        return true;
    }
    
    /** Deletes an item from the front of Deque. Return true if the operation is successful. */
    public boolean deleteFront() {
        if(this.isEmpty())
            return false;
        if(this.head == this.tail){
            this.head = this.tail = null;
        }
        else{
            this.head = this.head.next;
            this.head.prev = this.tail;
            this.tail.next = this.head;
        }
        this.count--;
        return true;
            
    }
    
    /** Deletes an item from the rear of Deque. Return true if the operation is successful. */
    public boolean deleteLast() {
        if(this.isEmpty())
            return false;
        if(this.head == this.tail){
            this.head = this.tail = null;
        }
        else{
            this.tail = this.tail.prev;
            this.tail.next = this.head;
            this.head.prev = this.tail;
        }
        this.count--;
        return true;
    }
    
    /** Get the front item from the deque. */
    public int getFront() {
        if(this.isEmpty())
            return -1;
        return this.head.value;
    }
    
    /** Get the last item from the deque. */
    public int getRear() {
        if(this.isEmpty())
            return -1;
        return this.tail.value;
    }
    
    /** Checks whether the circular deque is empty or not. */
    public boolean isEmpty() {
        return this.head == null;
    }
    
    /** Checks whether the circular deque is full or not. */
    public boolean isFull() {
        return this.count == this.length;
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque obj = new MyCircularDeque(k);
 * boolean param_1 = obj.insertFront(value);
 * boolean param_2 = obj.insertLast(value);
 * boolean param_3 = obj.deleteFront();
 * boolean param_4 = obj.deleteLast();
 * int param_5 = obj.getFront();
 * int param_6 = obj.getRear();
 * boolean param_7 = obj.isEmpty();
 * boolean param_8 = obj.isFull();
 */

File: deque/MyCircularDeque.py
'''

Runtime: 64 ms, faster than 94.68% of Python3 online submissions for Design Circular Deque.
Memory Usage: 14.5 MB, less than 63.03% of Python3 online submissions for Design Circular Deque.

'''
class MyCircularDeque:

    def __init__(self, k: int):
        """
        Initialize your data structure here. Set the size of the deque to be k.
        """
        self.length: int = k
        self.left = self.right = -1
        self.deque = [None] * self.length
        

    def insertFront(self, value: int) -> bool:
        """
        Adds an item at the front of Deque. Return True if the operation is successful.
        """
        if self.isFull():
            return False
        if self.left == -1:
            self.left = self.right = 0
            self.deque[self.left] = value
            return True
        elif self.left == 0:
            self.left = len(self.deque) - 1
            self.deque[self.left] = value
            return True
        else:
            self.left = self.left - 1
            self.deque[self.left] = value
            return True
        

    def insertLast(self, value: int) -> bool:
        """
        Adds an item at the rear of Deque. Return True if the operation is successful.
        """
        if self.isFull():
            return False
        if self.left == -1: 
            self.left = self.right = 0
            self.deque[self.left] = value
            return True
        elif self.right == self.length - 1:
            self.right = (self.right + 1) % self.length
            self.deque[self.right] = value
            return True
        else:
            self.right += 1
            self.deque[self.right] = value
            return True
        

    def deleteFront(self) -> bool:
        """
        Deletes an item from the front of Deque. Return True if the operation is successful.
        """
        if self.isEmpty():
            return False
        if self.left == self.right:
            self.left = self.right = -1
            return True
        
        else:
            if self.left == self.length - 1:
                self.left = 0
            else:
                self.left += 1
            return True
        
        

    def deleteLast(self) -> bool:
        """
        Deletes an item from the rear of Deque. Return True if the operation is successful.
        """
        if self.isEmpty():
            return False
        if self.left == self.right:
            self.left = self.right = -1
            return True
        
        else:
            if self.right == 0:
                self.right = self.length - 1
            else:
                self.right -= 1
            return True
        
        

    def getFront(self) -> int:
        """
        Get the front item from the deque.
        """
        if self.isEmpty():
            return -1
        return self.deque[self.left]
            

    def getRear(self) -> int:
        """
        Get the last item from the deque.
        """
        if self.isEmpty():
            return -1
        return self.deque[self.right]
        

    def isEmpty(self) -> bool:
        """
        Checks whether the circular deque is empty or not.
        """
        return self.left == -1
        

    def isFull(self) -> bool:
        """
        Checks whether the circular deque is full or not.
        """
        return (self.right + 1) % self.length == self.left


# Your MyCircularDeque object will be instantiated and called as such:
# obj = MyCircularDeque(k)
# param_1 = obj.insertFront(value)
# param_2 = obj.insertLast(value)
# param_3 = obj.deleteFront()
# param_4 = obj.deleteLast()
# param_5 = obj.getFront()
# param_6 = obj.getRear()
# param_7 = obj.isEmpty()
# param_8 = obj.isFull()